titus_isolate/allocate/forecast_ip_cpu_allocator.py

```diff
@@ -68,41 +70,73 @@ def __init__(self,
         self.__config_manager = config_manager
         self.__cnt_rebalance_calls = 0
         self.__event_log_manager = None
-
+        self.__call_meta = None # track things __place_threads call
+
+
+    def __with_report(func):
+        @wraps(func)
+        def wrapped(inst, cpu: Cpu, workload_id: str, workloads: dict, cpu_usage: dict, instance_id: str):
+            inst.__call_meta = {}
+            report = lambda cpu_ : inst.report_cpu_event(
+                inst.__event_log_manager,
+                cpu_,
+                list(workloads.values()),
+                cpu_usage,
+                instance_id,
+                inst.__call_meta) 
+            try:
+                cpu = func(inst, cpu, workload_id, workloads, cpu_usage, instance_id)
+                report(cpu)
+                return cpu
+            except Exception as e:
+                report(cpu)
+                raise e
+        return wrapped
+
+    @__with_report
     def assign_threads(self, cpu: Cpu, workload_id: str, workloads: dict, cpu_usage: dict, instance_id: str) -> Cpu:
+        self.__call_meta['type'] = 'add'
         curr_ids_per_workload = cpu.get_workload_ids_to_thread_ids()
+        return self.__place_threads(cpu, workload_id, workloads, curr_ids_per_workload, cpu_usage, True)
 
-        cpu = self.__place_threads(cpu, workload_id, workloads, curr_ids_per_workload, cpu_usage, True)
-        self.report_cpu_event(self.__event_log_manager, cpu, list(workloads.values()), cpu_usage, instance_id)
-        return cpu
-
+    @__with_report
     def free_threads(self, cpu: Cpu, workload_id: str, workloads: dict, cpu_usage: dict, instance_id: str) -> Cpu:
+        self.__call_meta['type'] = 'free'
         for t in cpu.get_claimed_threads():
             t.free(workload_id)
-        self.report_cpu_event(self.__event_log_manager, cpu, list(workloads.values()), cpu_usage, instance_id)
         return cpu
 
     def rebalance(self, cpu: Cpu, workloads: dict, cpu_usage: dict, instance_id: str) -> Cpu:
-        def __complete_rebalance():
-            self.report_cpu_event(self.__event_log_manager, cpu, list(workloads.values()), cpu_usage, instance_id)
-            return cpu
-
+        self.__call_meta = {'type': 'rebalance'}
```

Comment:
I wouldn't use this global variable that's getting reset and mutated.

Reply:
Maybe return more than one value from some methods?  e.g. `placement` and `meta_data`?